registration/registration.go

```diff
@@ -317,9 +371,46 @@ func (r *Registration) newRoundOpts() []newRoundOptionFunc {
 	}
 }
 
+func (r *Registration) verifyCert(
+	certificate *shared.OpaqueCert,
+	certPubKeyHint, nodeID []byte,
+) error {
+	var matchingKeys [][]byte
+
+	if len(certPubKeyHint) == 0 {
+		matchingKeys = append(matchingKeys, r.cfg.Certifier.PubKey.Bytes())
+	} else {
+		r.trustedKeysMtx.RLock()
+
+		certKeys := append(r.trustedPublicKeys, r.cfg.Certifier.PubKey.Bytes())
+		for _, certKey := range certKeys {
+			if len(certKey) >= len(certPubKeyHint) && bytes.Equal(certKey[:len(certPubKeyHint)], certPubKeyHint) {
+				matchingKeys = append(matchingKeys, certKey)
+			}
+		}
+
+		r.trustedKeysMtx.RUnlock()
+
+		if len(matchingKeys) == 0 {
+			return ErrNoMatchingCertPublicKeys
+		}
+	}
```

Comment:
If it can be avoided, try to write `if`s without `else`. The `append`s in the original code can be quite expensive, since appending data to a slice usually requires growing (and thereby copying) the slice first, also `certKeys` would then point to the same memory address as `r.trustedPublicKeys` which can lead to future errors if someone tries to modify `certKeys` and isn't aware that this isn't save to do.

My suggestion should reduce the amount copied data to a minimum and gets rid of the `else` block:

```suggestion
	matchingKeys := [][]byte{r.cfg.Certifier.PubKey.Bytes()}

	if len(certPubKeyHint) != 0 {
		r.trustedKeysMtx.RLock()
		matchingKeys = make([][]byte, 0)

		for _, certKey := range r.trustedPublicKeys {
			l := min(len(certKey), len(certPubKeyHint))
			if bytes.Equal(certKey[:l], certPubKeyHint) {
				matchingKeys = append(matchingKeys, certKey)
			}
		}
		l := min(len(r.cfg.Certifier.PubKey.Bytes()), len(certPubKeyHint))
		if bytes.Equal(r.cfg.Certifier.PubKey.Bytes()[:l], certPubKeyHint) {
			matchingKeys = append(matchingKeys, r.cfg.Certifier.PubKey.Bytes())
		}

		r.trustedKeysMtx.RUnlock()
		if len(matchingKeys) == 0 {
			return ErrNoMatchingCertPublicKeys
		}
	}
```

Reply:
IMHO, a map like `matchingKeys map[hint][][]byte` could be built when the keys are loaded instead of on every request.

Reply:
Sure, `trustedKeys` can even be replaced with such a map.

Reply:
Done